plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +487,70 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
+    def format_args(vars):
+        if isinstance(vars, str):
+            return '"{string}"'.format(string=vars)
+        elif isinstance(vars, bool):
+            if vars:
+                return 'true'
+            else:
+                return 'false'
+        return str(vars)
+
+    def process_complex_args(vars):
+        ret_out = []
+        if isinstance(vars, dict):
+            for k, v in vars.items():
+                if isinstance(v, dict):
+                    ret_out.append('{0}={{{1}}}'.format(k, process_complex_args(v)))
+                elif isinstance(v, list):
+                    ret_out.append("{0}={1}".format(k, process_complex_args(v)))
+                elif isinstance(v, (integer_types, float, str, bool)):
+                    ret_out.append('{0}={1}'.format(k, format_args(v)))
+                else:
+                    # only to handle anything unforeseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+        if isinstance(vars, list):
+            l_out = []
+            for item in vars:
+                if isinstance(item, dict):
+                    l_out.append("{{{0}}}".format(process_complex_args(item)))
+                elif isinstance(item, list):
+                    l_out.append("{0}".format(process_complex_args(item)))
+                elif isinstance(item, (str, integer_types, float, bool)):
+                    l_out.append(format_args(item))
+                else:
+                    # only to handle anything unforeseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+
+            ret_out.append("[{0}]".format(",".join(l_out)))
+        return ",".join(ret_out)
```

Comment:
If I may suggest:
```suggestion
    def process_complex_args(vars):
        def _value(item):
            if isinstance(item, (str, integer_types, float, bool)):
                return format_args(item)
            else:
                return process_complex_args(item)
            
        ret_out = []
        if isinstance(vars, dict):
            ret_out = ['{0}={1}'.format(k, _value(v)) for k, v in vars.items()]
            return "{" + ",".join(ret_out) + "}"

        elif isinstance(vars, list):
            ret_out = ['{0}'.format(_value(v)) for v in vars]
            return "[{0}]".format(",".join(ret_out))
            
        module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
```
I have not tested it as such, so beware of possible mistakes, but I think it should work fine, and it is simpler to read.

Reply:
Hey @russoz, that's surely a neat piece of code. I didn't test it, but the logic seem to be sound. I have few reservations though :). I started off with the function within  a function with recursion to solve a problem, and you're taking it further by function within a function within a function, and recursion...

Although I try to stay clear of list comprehensions as much as possible, more I read in to that segment, more I want to change it :), but after many tweaks, the code can succeed all the test cases, and the code aligns with the terrform structure for the CLI variables.

I'm adding integration testing to the module to test the variables, and I restructured the tests to be extensible to allow more tests to be added later on. Should we need to enhance the code, should be straightforward as the test bases are already covered.

Reply:
Fair enough.